Add WASI 0.3 launch blog (#153)
bytecodealliance/bytecodealliance.org: _posts/2026-06-11-WASI-0.3.md

```wit
// WASI 0.2
read-via-stream: func() -> result<input-stream, error-code>;
// WASI 0.3
read-via-stream: func() -> tuple<stream<u8>, future<result<_, error-code>>>;
```

## Changes to language bindings

One of the super powers of the component model is that it makes it trivial to
create bindings to and from other languages. With the addition of first-class
async it means guest binding generators can leverage that to create async
bindings that feel native to that language. Take for example the `wasi:http/handler`
interface. This interface exposes one function, `handle`, which is marked async:

```wit
interface handler {
  handle: async func(request: request) -> result<response, error-code>;
}
```

To implement an HTTP server in Rust with this, we can use the [wit-bindgen](https://github.com/bytecodealliance/wit-bindgen/tree/main#guest-rust)
crate. This maps the `interface handler` to a `trait Guest`, and maps the
`handle: async func` to an `async fn handle`:

```rust
use wasi::http::types::{ErrorCode, Request, Response};

impl Guest for Component {
    async fn handle(request: Request) -> Result<Response, ErrorCode> {
        // ...
    }
}
```

> Async support for guest binding generators is also in-progress for many languages including [Python](https://github.com/bytecodealliance/componentize-py), [JavaScript](https://github.com/dicej/componentize-js), [C#](https://github.com/bytecodealliance/wit-bindgen#guest-c), and [C](https://github.com/bytecodealliance/wit-bindgen/blob/main/crates/c/README.md). All of these languages however rely on _stackless coroutines_. But the Component Model's async ABI was designed from the ground up to accomodate both *stackful* and *stackless* coroutines side-by-side. An example of a language with those is Go. Instead of exposing async and non-async functions, Go’s runtime is able to convert synchronous-looking calls to async calls, and provides concurrent execution via virtual threads called “goroutines”.

Reviewer: ```suggestion
Async support for guest binding generators is also in-progress for many languages including [Python](https://github.com/bytecodealliance/componentize-py), [JavaScript](https://github.com/dicej/componentize-js), [C#](https://github.com/bytecodealliance/wit-bindgen#guest-c), and [C](https://github.com/bytecodealliance/wit-bindgen/blob/main/crates/c/README.md). All of these languages rely on _stackless coroutines_. But the Component Model's async ABI was designed from the ground up to accomodate both *stackful* and *stackless* coroutines side-by-side. An example of a language with those is Go. Instead of exposing async and non-async functions, Go’s runtime is able to convert synchronous-looking calls to async calls, and provides concurrent execution via virtual threads called “goroutines”.
```

The word "however" made me think there was some kind of catch to the first sentence, creating a subtle "uh oh, what's the problem" response; I'd just leave it out.

> Using [componentize-go](https://github.com/bytecodealliance/componentize-go/blob/2aa4ad80088ec18cd455cf9b5fecc0c9005fdffe/examples/wasip3/export_wasi_http_handler/handler.go) we can implement an HTTP server by exporting a `func Handle`. This  streaming bodies are produced from goroutines doing blocking calls. The runtime then parks the goroutine at the ABI boundary and resumes it when the stream is ready, without blocking the rest of the program:

Reviewer: ```suggestion
Using [componentize-go](https://github.com/bytecodealliance/componentize-go/blob/2aa4ad80088ec18cd455cf9b5fecc0c9005fdffe/examples/wasip3/export_wasi_http_handler/handler.go) we can implement an HTTP server by exporting a `func Handle`. This enables streaming bodies via goroutines that do blocking calls. The runtime then parks the goroutine at the ABI boundary and resumes it when the stream is ready, without blocking the rest of the program:
```

```wit
interface client { /* ... */ }
interface handler { /* ... */ }

// When used by guest bindings generators, grant the
// ability to make HTTP calls through the `client` import, and
// handle incoming HTTP requests through the `handler` export.
world service {
  import client;
  export handler;
}

// The middleware world is a super-set of the service world.
world middleware {
  include service; // ← Do everything that `service` can do.
  import handler;  // ← But also pass incoming requests down to another handler.
}
```

The `middleware` world replaces the 0.2-era `proxy` world, and is used to define
HTTP handlers which can forward requests to other handlers. What’s new in WASI
0.3 is that this can now perform [service
chaining](https://bytecodealliance.org/articles/how-wasm-components-enable-pluggable-middleware):
a pattern where components can be directly composed with one another. This means
components acting as microservices that frequently interop with other
microservices, do not need to go over the network. Instead a runtime can choose
to directly compose them with each other inside the same process. For most

> microservices this will change the time between mico-services from milliseconds to nanoseconds: six orders of magnitude.

Reviewer: ```suggestion
microservices this will change the time between microservices from milliseconds to nanoseconds: six orders of magnitude.
```

## Idiomatic async bindings

An interface like the `handler` interface in `wasi:http` lands in each guest language as that language's own async type.

```wit
interface handler {
  handle: async func(request: request) -> result<response, error-code>;
}
```

In Rust:

```rust
use wasi::http::types::{ErrorCode, Request, Response};

impl Guest for Component {
    async fn handle(request: Request) -> Result<Response, ErrorCode> {
        // ...
    }
}
```

In JavaScript:

```javascript
import { Response } from 'wasi:http/types';

export const handler = {
  async handle(request) {
    // ...
    return new Response(/* ... */);
  },
};
```

In Python:

```python
from wasi.http.types import Request, Response

class Handler:
    async def handle(self, request: Request) -> Response:
        # ...
```

> Rust, JavaScript, and Python all build on *stackless* coroutines, where the compiler rewrites an `async` function into a state machine. Go takes the other approach: a goroutine is a *stackful* coroutine that blocks on ordinary, synchronous-looking calls. The Component Model's async ABI accommodates both just as cleanly. With [componentize-go](https://github.com/bytecodealliance/componentize-go) and the [wit-bindgen](https://github.com/bytecodealliance/wit-bindgen) Go generator, the exported `handle` is a plain function that returns a value, and streaming bodies are produced from goroutines doing blocking calls. The runtime parks the goroutine at the ABI boundary and resumes it when the stream is ready, without blocking the rest of the program:

Reviewer: Nit: might want to link directly to https://github.com/bytecodealliance/wit-bindgen/tree/main/crates/go instead of just the top-level `wit-bindgen` repo.

---
title: "WASI 0.3 brings native async"
author: "Bailey Hayes"
github_name: "ricochet"
---

> **WASI 0.3 is official, and async is now native to WebAssembly Components.** The WASI Subgroup voted to ratify WASI 0.3.0, rebasing WASI onto the WebAssembly Component Model's async primitives. The 0.3.0 spec is stable, and runtime and toolchain support is landing now.

Reviewer 2: Thank you for updating the intro, this feels really good now!

> The work that `wasi:io` in WASIP2 used to do (pollables, input-streams, output-streams) has been pushed down into the canonical ABI, where the Component Model now offers these primitives natively.

Reviewer 2: We should link to what the canonical ABI is, most people reading this will probably not know. We refer to it 5 more times in the remainder of the post, which I think makes it worthwhile linking so people can find out more if they want to.

```suggestion
The work that `wasi:io` in WASIP2 used to do (pollables, input-streams, output-streams) has been pushed down into [the canonical ABI](https://github.com/WebAssembly/component-model/blob/main/design/mvp/CanonicalABI.md), where the Component Model now offers these primitives natively.
```

### The sandwich problem

In WASI 0.2, async is a `pollable` resource with the majority of the management of that resource in `wasi:io`. This worked well for cases where components are linked directly to the host or to one other component. There is only one `pollable`, owned, in many cases by the host, and the host runs the event loop.

It breaks the moment a component sits in the middle. Picture a three-layer sandwich:

```
  ┌─────────────────────────────┐
  │  Component A                │   awaits work
  │                             │
  │   pollable_B ───────────────┼──┐
  └─────────────────────────────┘  │  (A can only see B's pollable)
                                   │
  ┌─────────────────────────────┐  │
  │  Component B                │◀─┘
  │  (middle)                   │
  │                             │
  │   pollable_host ────────────┼──┐
  └─────────────────────────────┘  │  (the real readiness lives here)
                                   │
  ┌─────────────────────────────┐  │
  │  Host                       │◀─┘
  │                             │
  │  wake ⇒ pollable_host ready │
  └─────────────────────────────┘
```

> `pollable_host` is a resource inside B's instance. A cannot hold it, poll it, or name it. The best B can do is hand A its own `pollable_B` and promise to flip it when `pollable_host` fires, but there is no ABI hook for "when this pollable becomes ready, run this code in B." B would have to be scheduled and actively polling just to forward the wake-up. In practice the callback chain from host to B to A gets dropped on the floor, and the wake-up is lost in the middle of the sandwich.

Reviewer 2: The way I'd probably explain this is that in WASI 0.2 each component needed its own event loop/async runtime. That meant that individual components could be run on a host, but there was no way for those event loops to coordinate with one another.

WASI 0.3 makes it so the host is now the one in charge of managing the one event loop that is shared by all components.

Talking about "pollables" and "wakes" is not wrong, but it's a bit jargon-y, and I think we can probably simplify the explainer here a little.

Author: yeeted the diagram too.

This is why WASI 0.2 could express async components but not support native async composability.

### What the Component Model ABI gives us

The Component Model now has `stream<T>`, `future<T>`, and `async` as first-class constructs in the canonical ABI:

- `stream<T>` and `future<T>` are built-in to the canonical ABI. Like a resource, each is an owned handle, passing one across a component boundary transfers ownership from caller to callee, but unlike a resource it can't be borrowed, and the type itself is part of the Component Model rather than something a component imports or exports.
- The runtime, not each component, drives the scheduling. When a value has been delivered to a `future`, the runtime schedules whichever task is awaiting it, regardless of how many component boundaries sit between them. The writer that delivers that value might be the host, another component, or even the same component that holds the read end.
- Components export and import `async func`s directly, so the `start-foo` / `finish-foo` / `subscribe` three-step from WASI 0.2 disappears.

> - The model is *completion-based*, not *readiness-based*. You hand the runtime an operation and await its completion, the way Linux's `io_uring` or Windows' I/O Completion Ports work — rather than asking "is this ready yet?" and then doing the I/O yourself, the way POSIX `poll(2)` does. A readiness-style API can still be emulated on top with minimal overhead when a guest needs it.

Reviewer 2: This first sentence here is a bit of a run-on. I also feel like people are generally aware enough that "io_uring good" that we don't need to explain in detail what the differences are. I'd re-phrase it like this:

```suggestion
- The async model is *completion-based*, not *readiness-based*. This is similar to the ultra-efficient Linux `io_uring` and Windows' IOCP/`IoRing` APIs. An `epoll`/`kqueue`-style readiness API can be emulated on top of this for programs which need the compatibility.
```

The middle component in our sandwich now passes the future through. There is no wake-up to forward and nothing for B to plumb by hand.

## What this looks like in the WIT

> Most of the per-package change is a mechanical consequence of the rebase: replace `pollable` with `future<T>`, replace `input-stream` / `output-stream` with `stream<u8>`, and collapse `start-foo` / `finish-foo` pairs into a single `async func`. A few cross-cutting translations before we walk the packages:

Reviewer 2: We have multiple changes per package, so I think we should pluralize it? I feel like we can also shorten the text a little by cheating and using shorthand:

```suggestion
Most of the per-package changes are a mechanical consequence of the rebase: replace `pollable` with `future<T>`, replace `{input,output}-stream` with `stream<u8>`, and unify `{start,finish}-foo` pairs into single `async func`s. A few cross-cutting translations before we walk the packages:
```

Reviewer 2: Actually: I feel like this reads a little AI-y. We're essentially listing things twice here: once in a sentence, and once in a table. We should probably just keep the table and change the leading sentence.

 
```suggestion
Most of the changes in the 0.3 interfaces are entirely mechanical. WASI 0.2 had to perform some acrobatics to make async work, but now that async is native to the component model we can write the same things we did before but much more ergonomically. Here is a overview of the patterns we were encoding in WASI 0.2 with the `wasi:io` package, and what those patterns now look like in 0.3 with Component Model async:
```

| WASI 0.2 (`wasi:io`)       | WASI 0.3 (Component Model)             |
| -------------------------- | -------------------------------------- |
| `resource pollable`        | `future<T>`                            |
| `resource input-stream`    | `stream<u8>`                           |
| `resource output-stream`   | `stream<u8>` (written-to direction)    |
| `poll(list<pollable>)`     | `await` on a future (runtime-handled)  |
| `subscribe()` on resource  | return a `future<...>` from the call   |
| `start-foo` / `finish-foo` | `foo: async func(...)`                 |

One more pattern to know about before we dive into specific packages. The common "give me a stream for this resource" call now returns both a stream and a completion future:

```wit
// WASI 0.2
read-via-stream: func() -> result<input-stream, error-code>;

// WASI 0.3
read-via-stream: func() -> tuple<stream<u8>, future<result<_, error-code>>>;
```

> WASIP2 did surface terminal errors, but inline with each read call on the stream (via the stream-error variant). That meant callers only learned the outcome if they kept reading. Stop early and you never find out if the stream errored or just closed. In WASIP3 the future resolves independently of how much of the stream you consume, and it's a first-class async value you can await, forward across components, or compose with other futures.

Reviewer 2: ````suggestion
A problem with WASI 0.2 was that it surfaced terminal errors inline on each read call on the stream. That meant callers only learned the outcome if they kept reading. If readers stopped early they could not distinguish between a stream close and an error.

In WASI 0.3 streams now return an additional future which resolves independently of how much of the stream is consumed, solving the stream status problem of WASI 0.2:

```wit
// WASI 0.2
read-via-stream: func() -> result<input-stream, error-code>;

// WASI 0.3
read-via-stream: func() -> tuple<stream<u8>, future<result<_, error-code>>>;
```
````

### `wasi:io` — removed

> This package is no longer needed. Every consumer uses the Component Model's native types.

Reviewer 2: If we land the changes above, we can remove this.
